cartographer_ros/cartographer_ros/rosbag_wtf_main.cc

```diff
@@ -0,0 +1,121 @@
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+#include "nav_msgs/Odometry.h"
+#include "ros/ros.h"
+#include "ros/time.h"
+#include "rosbag/bag.h"
+#include "rosbag/view.h"
+#include "sensor_msgs/Imu.h"
+#include "sensor_msgs/LaserScan.h"
+#include "sensor_msgs/MultiEchoLaserScan.h"
+#include "sensor_msgs/PointCloud2.h"
+#include "tf2_eigen/tf2_eigen.h"
+#include "tf2_msgs/TFMessage.h"
+#include "tf2_ros/buffer.h"
+#include "urdf/model.h"
+
+DEFINE_string(bag_filename, "", "Bag to process");
+
+namespace cartographer_ros {
+namespace {
+
+struct TimestampData {
+  ros::Time last_timestamp;
+  ::cartographer::common::Histogram histogram;
+};
+
+void Run(const string& bag_filename) {
+  rosbag::Bag bag;
+  bag.open(bag_filename, rosbag::bagmode::Read);
+  rosbag::View view(bag);
+  const ::ros::Time begin_time = view.getBeginTime();
+  const double duration_in_seconds = (view.getEndTime() - begin_time).toSec();
+
+  std::map<string, TimestampData> timestamp_data;
+
+  for (const rosbag::MessageInstance& message : view) {
+    string frame_id;
+    ros::Time time;
+    if (message.isType<sensor_msgs::PointCloud2>()) {
+      auto msg = message.instantiate<sensor_msgs::PointCloud2>();
+      time = msg->header.stamp;
+      frame_id = msg->header.frame_id;
+    } else if (message.isType<sensor_msgs::MultiEchoLaserScan>()) {
+      auto msg = message.instantiate<sensor_msgs::MultiEchoLaserScan>();
+      time = msg->header.stamp;
+      frame_id = msg->header.frame_id;
+    } else if (message.isType<sensor_msgs::LaserScan>()) {
+      auto msg = message.instantiate<sensor_msgs::LaserScan>();
+      time = msg->header.stamp;
+      frame_id = msg->header.frame_id;
+    } else if (message.isType<sensor_msgs::Imu>()) {
+      auto msg = message.instantiate<sensor_msgs::Imu>();
+      time = msg->header.stamp;
+      frame_id = msg->header.frame_id;
+    } else if (message.isType<nav_msgs::Odometry>()) {
+      auto msg = message.instantiate<nav_msgs::Odometry>();
+      time = msg->header.stamp;
+      frame_id = msg->header.frame_id;
+    } else {
+      continue;
+    }
+
+    if (!timestamp_data.count(frame_id)) {
+      timestamp_data.emplace(
+          frame_id, TimestampData{time, ::cartographer::common::Histogram()});
+    } else {
+      auto& entry = timestamp_data.at(frame_id);
+      const double delta_t_sec = (time - entry.last_timestamp).toSec();
+      if (delta_t_sec < 0) {
+        LOG(ERROR) << "Sensor with frame_id \"" << frame_id
+                   << "\" jumps backwards in time. Make sure that the bag "
+                      "contains the data for each frame_id sorted by "
+                      "header.stamp, i.e. the order in which they were "
+                      "acquired from the sensor.";
+      }
+      entry.histogram.Add(delta_t_sec);
+      entry.last_timestamp = time;
+    }
+  }
+  bag.close();
+
+  constexpr int kNumBucketsForHistogram = 10;
+  for (const auto& entry_pair : timestamp_data) {
```

Comment:
Looks like this will output every timestamp even if everything is fine, but I guess this is intended for this first version. 
(My understanding was the output should be short enough to paste it into an issue.)

Reply:
I do not fully understand what you are asking. It outputs an ASCII histogram of each sensors timestamp deltas. This is still short enough to be pasted into an issue, but you are right that it might get longer over time.

I changed the issue template to ask for a gist of the output instead.